internal/http/services/owncloud/ocdav/propfind/propfind.go

```diff
@@ -255,35 +248,122 @@ func (p *Handler) HandleSpacesPropfind(w http.ResponseWriter, r *http.Request, s
 		return
 	}
 
-	// retrieve a specific storage space
-	space, rpcStatus, err := spacelookup.LookUpStorageSpaceByID(ctx, client, spaceID)
+	ref, err := spacelookup.MakeStorageSpaceReference(spaceID, r.URL.Path)
 	if err != nil {
-		sublog.Error().Err(err).Msg("error looking up the space by id")
-		w.WriteHeader(http.StatusInternalServerError)
+		sublog.Debug().Msg("invalid space id")
+		w.WriteHeader(http.StatusBadRequest)
+		m := fmt.Sprintf("Invalid space id: %v", spaceID)
+		b, err := errors.Marshal(http.StatusBadRequest, m, "")
+		errors.HandleWebdavError(&sublog, w, b, err)
 		return
 	}
 
-	if rpcStatus.Code != rpc.Code_CODE_OK {
-		errors.HandleErrorStatus(&sublog, w, rpcStatus)
+	client, err := p.getClient()
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
-	resourceInfos, sendTusHeaders, ok := p.getResourceInfos(ctx, w, r, pf, []*provider.StorageSpace{space}, r.URL.Path, true, sublog)
-	if !ok {
-		// getResourceInfos handles responses in case of an error so we can just return here.
+	metadataKeys, _ := metadataKeys(pf)
+
+	// stat the reference and request the space in the field mask
+	res, err := client.Stat(ctx, &provider.StatRequest{
+		Ref:                   &ref,
+		ArbitraryMetadataKeys: metadataKeys,
+		FieldMask:             &fieldmaskpb.FieldMask{Paths: []string{"*"}}, // TODO use more sophisticated filter? we don't need all space properties, afaict only the spacetype
+	})
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
+	if res.Status.Code != rpc.Code_CODE_OK {
+		status := rstatus.HTTPStatusFromCode(res.Status.Code)
+		if res.Status.Code == rpc.Code_CODE_ABORTED {
+			// aborted is used for etag an lock mismatches, which translates to 412
+			// in case a real Conflict response is needed, the calling code needs to send the header
+			status = http.StatusPreconditionFailed
+		}
+		m := res.Status.Message
+		if res.Status.Code == rpc.Code_CODE_PERMISSION_DENIED {
+			// check if user has access to resource
+			sRes, err := client.Stat(ctx, &provider.StatRequest{Ref: &provider.Reference{ResourceId: ref.GetResourceId()}})
```

Comment:
What is the difference between this stat and the original one above?

Reply:
The original stat might return an error, to check if we need to hide that error this stat checks if we are allowed to stat the space (well only the resource id, ignoring any path). If this stat returns OK we have read access and can return other errors like PrecondtitionFailed when an etag does not match.

The more I think about it, the more I think we should move 'hiding the existence of spaces' int the storage drivers. an id based lookup has to find the space first. if a user has no access to it we can always return 404 and stop any further request processing.

But that is a task for a separate PR.

Reply:
I have to admit this is really confusing. The testsuite does not even cover all requests and we actually need to do this error code handling for every request type...

I'll move this to the driver. That will also allow eos to return 403 forbidden and decomposedfs to return 404 not found. And we can even configure this kind of hiding per space.